packages/did/src/Did.chain.ts

```diff
@@ -105,48 +149,120 @@ function decodeDidChainRecord(
-  return queryById(didId)
+  const { identifier, fragment } = parseDidUrl(didUri)
+  if (fragment) {
+    throw new Error(`The provided URI ${didUri} must not contain any fragment.`)
+  }
+  return queryById(identifier)
 }
 
-export async function queryKey(
-  did: string,
-  keyId: string
+// Query a given key given the DID identifier and the key ID.
+export async function queryDidKey(
+  keyUri: IDidKeyDetails['id']
 ): Promise<IDidKeyDetails | null> {
-  const encoded = await queryEncoded(getIdentifierFromKiltDid(did))
-  if (encoded.isNone) return null
-  const keyIdU8a = Crypto.coToUInt8(keyId)
-  let key: IDidKeyDetails | null = null
-  encoded.unwrap().publicKeys.forEach((keyDetails, id) => {
-    if (id.eq(keyIdU8a)) {
-      key = decodeDidPublicKeyDetails(did, id, keyDetails)
-    }
-  })
-  return key
+  const { identifier, fragment } = parseDidUrl(keyUri)
+  if (!fragment) {
+    throw new Error(
+      `The provided URI ${keyUri} does not contain a valid fragment for key ID.`
+    )
+  }
+  const didDetails = await queryById(identifier)
+  if (!didDetails) {
+    return null
+  }
+  return (
+    didDetails.publicKeys
+      .filter((key) => {
+        return key.id === keyUri
+      })
+      .pop() || null
+  )
 }
 
-export async function queryLastTxIndex(did: string): Promise<BN> {
-  const identifier = getIdentifierFromKiltDid(did)
-  const encoded = await queryEncoded(identifier)
-  if (encoded.isNone) return new BN(0)
-  return encoded.unwrap().lastTxCounter.toBn()
+export async function queryServiceEndpoints(
+  didUri: IDidDetails['did']
+): Promise<IDidServiceEndpoint[]> {
+  const { identifier, fragment } = parseDidUrl(didUri)
+  if (fragment) {
+    throw new Error(`The provided URI ${didUri} must not contain any fragment.`)
+  }
+  const encoded = await queryAllServicesEncoded(identifier)
+  return encoded.map((e) => decodeServiceChainRecord(e, didUri))
+}
+
+export async function queryServiceEndpoint(
+  serviceUri: IDidServiceEndpoint['id']
+): Promise<IDidServiceEndpoint | null> {
+  const { identifier, fragment } = parseDidUrl(serviceUri)
+  if (!fragment) {
+    throw new Error(
+      `The provided URI ${serviceUri} does not contain a valid fragment for service ID.`
+    )
+  }
+  const serviceEncoded = await queryServiceEncoded(identifier, fragment)
+  if (serviceEncoded.isNone) return null
+
+  return decodeServiceChainRecord(
+    serviceEncoded.unwrap(),
+    getKiltDidFromIdentifier(identifier, 'full')
+  )
+}
+
+export async function queryEndpointsCounts(
+  didUri: IDidDetails['did']
+): Promise<number> {
+  const { identifier, fragment } = parseDidUrl(didUri)
+  if (fragment) {
+    throw new Error(`The provided URI ${didUri} must not contain any fragment.`)
+  }
+  const blockchain = await BlockchainApiConnection.getConnectionOrConnect()
+  const count = await blockchain.api.query.did.didEndpointsCount<u32>(
+    identifier
+  )
+  return count.toNumber()
+}
+
+export async function queryLastTxIndex(
```

Comment:
Since you are querying the `txIndex` the `txCounter` should be renamed to `txIndex` or vis-à-vis. It would be better for developer experience.

Reply:
https://github.com/KILTprotocol/sdk-js/pull/428/commits/abb47ef6164088db1e0ce067853180b11d58a26c